src/com/georgidinov/algorithms/sorting/bubblesort/BubbleSort.java

```diff
@@ -42,40 +39,37 @@ private void bubbleSortImplOne(Integer[] array) {
         }
     }
 
-    private void bubbleSortImplTwo(Integer[] array) {
+    private void bubbleSortStreamImpl(Integer[] array) {
         int first = 0;
         int last = array.length - 1;
-        IntStream
-                .range(first, last)
-                .flatMap(index -> {
-                    MyUtil.printInColor("index=" + index + ", last=" + (last - index), ThreadColor.ANSI_BLUE);
-                    return IntStream.range(first, last - index);
-                })
+        IntStream.range(first, last)
+                .flatMap(index -> IntStream.range(first, last - index))
                 .forEach(current -> {
                     int next = current + 1;
-                    System.out.println("current=" + array[current] + ", next=" + array[next]);
                     if (array[current] > array[next]) {
-                        swap(array, (current), next);
+                        swap(array, current, next);
                     }
                 });
     }
 
-    private void bubbleSortImplThree(Integer[] array) {
-        int i = 0, n = array.length;
+    //optimized
+    private void bubbleSortFlagImpl(Integer[] array) {
+        int first = 0;
+        int last = array.length - 1;
 
-        boolean swapNeeded = true;
-        while (i < n - 1 && swapNeeded) {
-            swapNeeded = false;
-            for (int j = 1; j < n - i; j++) {
-                if (array[j - 1] > array[j]) {
-                    swap(array, (j - 1), j);
-                    swapNeeded = true;
+        while (first < last) {
+            boolean isSwapNeeded = false;
+            for (int current = 0; current < (last - first); current++) {
+                int next = current + 1;
+                if (array[current] > array[next]) {
+                    swap(array, current, next);
+                    isSwapNeeded = true;
```

Comment:
If swap was needed, the array is not sorted, and set the flag to false.

Reply:
Resolved.